src/themis/secure_keygen.h

```diff
@@ -58,49 +120,84 @@ themis_status_t themis_gen_rsa_key_pair(uint8_t* private_key,
                                         size_t* public_key_length);
 
 /**
- * @brief generate EC key pair
- * @param [out] private_key buffer for private key to store. May be set to NULL for private key
- * length determination
- * @param [in, out] private_key_length length of private_key
- * @param [out] public_key buffer for public key to store. May be set to NULL for public key length
- * determination
- * @param [in, out] public_key_length length of public key
- * @return THEMIS_SUCCESS on success or THEMIS_FAIL on failure
- * @note If private_key==NULL or public_key==NULL or private_key_length is not enought for private
- * key storage or public_key_length is not enought for public key storage then
- * THEMIS_BUFFER_TOO_SMALL will return and private_key_length and public_key_length will store
- * lengths of buffers needed for private key and public key store respectively
+ * Generates an EC key pair.
+ *
+ * @param [out]     private_key         buffer for private key
+ * @param [in,out]  private_key_length  length of private key in bytes
+ * @param [out]     public_key          buffer for public key
+ * @param [in,out]  public_key_length   length of public key in bytes
+ *
+ * New EC key pair is generated and written into `private_key` and
+ * `public_key` buffers which must have at least `private_key_length`
+ * and `public_key_length` bytes available respectively. Note that
+ * length parameters are _pointers_ to actual length values.
+ *
+ * You can pass NULL for `private_key` and `public_key` in order to
+ * determine appropriate buffer length. In this case the lengths are
+ * written into provided locations, no key data is generated, and
+ * THEMIS_BUFFER_TOO_SMALL is returned.
+ *
+ * @returns THEMIS_SUCCESS if the keys have been generated successfully
+ * and written to `private_key` and `public_key`.
+ *
+ * @returns THEMIS_BUFFER_TOO_SMALL if the key lengths have been written
+ * to `private_key_length` and `public_key_length`.
+ *
+ * @exception THEMIS_FAIL if key generation has failed.
+ *
+ * @exception THEMIS_INVALID_PARAM if `private_key_length` or
+ * `public_key_length` is NULL.
+ *
+ * @exception THEMIS_BUFFER_TOO_SMALL if `private_key` and `public_key`
+ * are not NULL, but `private_key_length` or `public_key_length` in not
+ * sufficient to hold a generated key.
  */
 THEMIS_API
 themis_status_t themis_gen_ec_key_pair(uint8_t* private_key,
                                        size_t* private_key_length,
                                        uint8_t* public_key,
                                        size_t* public_key_length);
 
-enum themis_key_kind {
+/**
+ * Kind of an asymmetric Themis key.
+ */
+typedef enum themis_key_kind {
+    /** Invalid key buffer. */
     THEMIS_KEY_INVALID,
+    /** Private RSA key. */
```

Comment:
for what these comment lines which duplicate enum values?

Reply:
> for what these comment lines which duplicate enum values?

Doxygen output looks better this way. If we describe enumerators then the enumeration gets a 'detailed description' section like all other declarations.

Though yeah, it _does feel_ like infamous Java getter docs:

```java
/**
 * Returns success value.
 *
 * Use this method to get success value as boolean.
 *
 * @return boolean value `true` if successful, `false` if not successful.
 */
public boolean getSuccess() {
    return succes;
}
```

🙄